### PR TITLE
Add FLAG_SECURE to prevent screenshots (closes #9)

### LINKED ISSUE #9: [BSI][0.5.4-alpha] App Snapshots Enabled
Rating: Low
 
Description:
When an app is sent to the background, Android takes a snapshot of the latest view of the application and saves it on the file system. If this snapshot contains sensitive data, a malicious app can obtain such data from the sanpshot.

The snapshot is saved outside of the app's sandbox, therefore any malicious app can access it only if it has root privileges or can bypass sandboxing limitations.

Proof of concept - see picture below:
android_background_snapshot

### PR DESCRIPTION
**Pull request by m1ga**
## Checklist

* [x] Test your changes as thoroughly as possible before you commit them. Preferably, automate your test by unit/integration tests.
* [x] If this PR comes from a fork, please [Allow edits from maintainers](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
* [x] Set a speaking title. Format: {task_name} (closes #{issue_number}). For example: Use logger (closes #41)
* [x] [Link your Pull Request to an issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) (if applicable)
* [ ] Create Work In Progress [WIP] pull requests only if you need clarification or an explicit review before you can continue your work item.
* [x] Make sure that your PR is not introducing _unncessary_ reformatting (e.g., introduced by on-save hooks in your IDE)
* [x] Check our [Contribution Guidelines](https://github.com/corona-warn-app/cwa-app-android/blob/master/CONTRIBUTING.md)

## Description

Adding FLAG_SECURE to block screenshots from being made:

Before:
screen_0

After:
screen_1

Fixes https://github.com/corona-warn-app/cwa-app-android/issues/9

### CODE REVIEW

**Review by jakobmoellerdev — COMMENTED** [review]
Looks good to me, will test and then approve.

**Comment by uschindler** [comment]
Disallowing screenshots in an app globally is preventing users from documenting their status. It should be allowed for anybody to visualize hisher health status (if they want this) and save it as Screenshot. Most modern banking apps also allow this, unless it's screens with sensitive data (like TAN input screens).

Please only disallow screenshots from activities that are critical for security, like TAN input screen.

**Comment by m1ga** [comment]
@uschindler good point! I removed it globally and just moved it to the TAN input screen

**Comment by uschindler** [comment]
Thanks! I can't speak for BSI or the project team here, but this app should really be as open as possible. I am sure many people want to share screenshots.

**Comment by elsweppo** [comment]
Your commit now prevents the user from taking screenshots in the TAN input screen, but as soon as the user goes now further to other screens/fragments the setting of disallowing screenshots is still set. To avoid this behaviour and limit the the disabling screenshot feature only to the TAN input screen, you have to remove the option FLAG_SECURE as soon as the user will leave the screen. 
So maybe it would be better, to add the flag on onResume and clear the flag again on onStop.

**Comment by m1ga** [comment]
@elsweppo Thank you for the hint! Can confirm that it was blocked before this last change and it is unblocked (so only the TAN screen is blocked) when closing it again

**Comment by uschindler** [comment]
The close of this PR seems to be a mistake introduced by the issue number as an example in the default template for new pull requests. As #57 was merged and this PR is wrongly referenced, it was closed.

Please reopen.

**Comment by m1ga** [comment]
and it is mentioned everywhere which is very confusing  :smile:

**Comment by uschindler** [comment]
But 41 is not 42, the answer to everything! 🤣

**Comment by tkowark** [comment]
sorry about that. There is already a PR to fix the Template...

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/SubmissionTanFragment.kt [hunk 1]
```diff
@@ -28,6 +29,16 @@ class SubmissionTanFragment : BaseFragment() {
         return binding.root
     }
 
+    override fun onResume() {
+        super.onResume()
+        activity?.window?.setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE)
```
Comment (kedar1607) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/SubmissionTanFragment.kt:34]: Does it makes sense to do this based on flavors? e.g. we can still keep screen shots enabled for debug mode and disable them for release mode. This way communication with the design team can be easier while developing this app.

Comment (uschindler) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/SubmissionTanFragment.kt:34]: I'd like to have screen shots for the non-privacy related stuff enabled, but disallow it for the TAN input and several other submission fragments.
I agree, in debug mode, screenshos should be enabled anywhere.

The main problem I have is the "hack" to enable or disable screenshots based on fragment. I will now open another issue to split the main activity (the app currently already has 2 activities: onboarding and "main") from the submission one. Then you can declare the "submission" activity's window as secured, while the main activity and the onboarding activity are not specifically secured. Of course a debug mode would really be good.

Comment (uschindler) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/SubmissionTanFragment.kt:34]: I opened #77

**Comment by uschindler** [comment]
Hi,
the current implementation with screenshotting/app snapshots looks very hacky:  We manipulate the activity's window on enable/disable of the fragment.

As a quick hack this PR seems fine, but I tend to split the submission (which is very privacy related) from the default main activity. The onboarding activity is already separate, so should be the submission. Each activity has its own window, so the correct way to handle this issue is to disable screenshotting for the whole submission activity (if it would be separate): see #77 with details and rationale.

**Comment by m1ga** [comment]
I'll close this one since #108 was merged and there is still an open issue to decide what to do: https://github.com/corona-warn-app/cwa-app-android/issues/77
If it should only be a specific activity/fragment a new PR can be created.

ps: and finally number 41 is closed :smile:

**Comment by marcmuschko** [comment]
Dear @m1ga,

thank you for the proposed idea regarding the screenshot issue, we really do appreciate the work and thoughts that went into this PR. 
We took the proposal and discussion seriously. It has been clarified internally and with the project stakeholders. As several streams partake in a decision like this, especially when security and privacy are at stake, it is, due to very strict time limitations, not always possible to find a satisfying compromise for all parties involved. 
For this version it was decided to prevent in-app screenshots for all screens.

As this is a general product feature and not specific to Android only, please feel free to create an issue in the documentation repository, as this might be up for discussion again for a later version and we will close the related BSI issue shortly.

Nevertheless, thank you very much for the proposal! 
Best,

Hee Tatt & Marc

**Comment by ChristianMertes** [comment]
> thank you for the proposed idea regarding the screenshot issue, we really do appreciate the work and thoughts that went into this PR.
> We took the proposal and discussion seriously. It has been clarified internally and with the project stakeholders. As several streams partake in a decision like this, especially when security and privacy are at stake, it is, due to very strict time limitations, not always possible to find a satisfying compromise for all parties involved.

This explanation makes absolutely zero sense.

**Pull request closed without merging**